docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,208 @@
+this behavior and disable it. The only solution is to not allow scripts to load from any vulnerable domain, and to host
+them yourself instead.
+
+**The bigger your website, the harder this is to deal with.**
+
+If you have a small site and immediate control over the codebase, this could be an easy problem to patch.
+
+For large sites at big companies that load analytics payloads, third party auth, and other scripts from these vulnerable
+domains, it can be difficult to mitigate. Heck, a lot of big companies have legacy JSONP endpoints on the _same domain
+as the site._  There's not much you can do about that.
+
+### File Upload Bypass
+
+Let's say your CSP is locked to only allow scripts from your domain.
+
+If your site allows file uploads and downloads and doesn't discriminate against files that end in `.js` with content
+type `application/javascript`. A user can simply upload a file `malicious-code.js`, and load it as a script tag,
+directly from your domain.  
+
+```html
+
+<script src=www.yourdomain.com/myfiles/malicious-code.js></script>
+```
+
+You have to be very careful about what your users can upload.  In some cases the browser will try to automatically detect
+and fix MIME types in a process called sniffing, which could convert something you didn't block back into `application/javascript` during download.
+File downloads need to [include the header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options) `X-Content-Type-Options: nosniff` to prevent that.  
+
+As you can see, these are nuanced details that are easy to miss.
+
+### Other vulnerabilities
+
+These are just a few of the potential ways to bypass the CSP's protections. The
+list [goes on](https://book.hacktricks.xyz/pentesting-web/content-security-policy-csp-bypass).
+
+### Mitigations
+
+For large sites, there may be just too many gaps in the armor to treat the CSP as a functional line of defense, and
+many big sites have disabled it entirely.
+[This paper](https://research.google/pubs/pub45542/) found that of those that use it **over 94% of sites have
+ineffective CSPs!**
+
+Long story short, there's a good chance your CSP isn't doing much to protect your site from XSS.
+
+Nonce-based CSP strategies can help mitigate the above attacks, where the server adds a random nonce attribute to every
+script tag and defines in the CSP. Embedding a script tag smart enough to clone the nonce isn't possible without
+JavaScript.
+
+Unfortunately this strategy is the most work to implement and is rarely used.  Getting it to work with CDNs and caching can be a challenge.
+
+```text title=CSP
+script-src 'nonce-randomString';
+```
+
+```html title="Script Tag"
+<script nonce="randomString" ...
+```
+
+Defining a nonce and a domain `script-src` together means that _either_ the domain or the nonce can be used, which seems
+like a gotcha that could catch many users.
```

Comment:
I would throw this in here https://csp-evaluator.withgoogle.com/ as a call to action: "see for yourself how secure your site's CSP policy is"

Reply:
Also you could add "For an idea how effective CSP is, https://google.com has no CSP headers set. https://mail.google.com is a prime example of what a strict csp policy looks like, it simply uses a nonce to get all the benefits of a strong CSP policy."